### PR TITLE
Contextually type implemented properties

### LINKED ISSUE #1373: Consider contextually typing class member functions by their base class/interface members
here is my code

```ts
    export enum SortDir
    {
        Asc,
        Desc,
        Flip
    }

    export interface ISortInfo
    {
        getvalue(x: QuerySummary): any; //change to number|string in TypeScript 1.4
        order: SortDir;
        ordercalc?: number;
    }

    class SortByQueryId implements ISortInfo
    {
        getvalue = x => x.QueryId;
        order = SortDir.Flip;
    }
```

when I type "x." I expect to see the properties of QuerySummary but I do not. VS still thinks the type is 'any'

### LINKED ISSUE #3667: Contextually type class property initializers by `extends`/`implements` types
When a property has exactly one type from an `extends` or `implements` clause, we should contextually type that property's initializer by the type from the clause. For example:

```ts
interface ThingListener {
	handleEvent: (x: MouseEvent) => void;
}

class Foo implements ThingListener {
	handleEvent = x => {
		// No error, but this code is incorrect
		console.log(x.timestamp);
	}
}
```

Ideally we could take this one step further when widening. The following behavior is also undesirable:
```ts
interface HasLength {
	length: number
}

class Foo implements HasLength {
	// length: any, not really what we intended
	length = undefined;
}

var x = new Foo();
x.length = 'wat'; // should not be allowed
```

e.g. #3666

### PR DESCRIPTION
**Pull request by sandersn**
Fixes #3667. Fixes #1373.

```ts
interface I {
  f(n: number): string;
  p: string;
}
class C implements I {
  f = n => n.toString(); // n is correctly typed as `number`
  p = undefined; // correctly retains type `string`
}
```

Interesting baseline changes in the latest commit:

```ts
let x: number;
let y: string;
x = y = undefined;
```

is now illegal. This is a breaking change.

```ts
let [...r] = null
```

This is now legal because null gets contextually typed as an array, although node says "Cannot read property 'Symbol(Symbol.iterator)' of null" when running the code.

There are also a couple of implicit any errors that went missing from wideningTuples3 and wideningTuples5. I still need to find out why the error no longer fires:

```ts
var a: [any];
var b = a = [undefined, null];
```

Should have an an error when b implicitly gets the type `[any, any]`.

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -2895,6 +2902,34 @@ namespace ts {
             return unknownType;
         }
 
+        function getTypeOfBasePropertyDeclaration(declaration: VariableLikeDeclaration) {
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: change to `declaration: PropertyDeclaration`.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -2895,6 +2902,34 @@ namespace ts {
...
+                    return getTypeOfSymbol(property);
+                }
+            }
+        }
+
+        function getPropertyFromBaseInterfaces(declaration: ClassLikeDeclaration, propertyName: string): Symbol {
+            const implementedTypeNodes = getClassImplementsHeritageClauseElements(declaration);
+            if (implementedTypeNodes) {
+                for (const typeRefNode of implementedTypeNodes) {
+                    const t = getTypeFromTypeReference(typeRefNode);
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: This isn't going to work properly for generic types, this gets you the uninstantiated base type.

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -2895,6 +2902,34 @@ namespace ts {
...
+                    if (t !== unknownType) {
+                        for (const property of getPropertiesOfType(t)) {
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: Why the loop when you could use `getPropertyOfType` or `getPropertyOfObjectType`?

File: src/compiler/checker.ts [hunk 4]
```diff
@@ -2895,6 +2902,34 @@ namespace ts {
...
+        function getTypeOfBasePropertyDeclaration(declaration: VariableLikeDeclaration) {
+            if (declaration.parent.kind === SyntaxKind.ClassDeclaration) {
+                const property = getPropertyFromBaseInterfaces(<ClassLikeDeclaration>declaration.parent, declaration.symbol.name);
+                if (property) {
+                    return getTypeOfSymbol(property);
+                }
+            }
+        }
+
+        function getPropertyFromBaseInterfaces(declaration: ClassLikeDeclaration, propertyName: string): Symbol {
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: Why only base interfaces? Shouldn't we look at the base class for a contextual type?

File: src/compiler/checker.ts (hunk 2 above)
Comment (ahejlsberg) [inline src/compiler/checker.ts]: Actually, nevermind about the uninstantiated base type.

File: tests/baselines/reference/bestCommonTypeOfTuple2.types [hunk 5]
```diff
@@ -29,7 +29,7 @@ class F extends C { f }
 class C1 implements base1 { i = "foo"; c }
 >C1 : C1
 >base1 : base1
->i : string
+>i : any
```
Comment (DanielRosenwasser) [inline tests/baselines/reference/bestCommonTypeOfTuple2.types]: This is a regression. `"foo"` should be contextually typed with `any`, but since that doesn't affect the type of `"foo"`, `i` should still infer its type as `string`.

File: src/compiler/checker.ts (hunk 3 above)
Comment (sandersn) [inline src/compiler/checker.ts]: I didn't know about those. Thanks.

File: src/compiler/checker.ts (hunk 4 above)
Comment (sandersn) [inline src/compiler/checker.ts]: I thought that a derived class property would override the base class property (making their types unrelated), but on playing with it in the playground, it looks like that's not the case -- it has to be a subclass of the base class' property's type. I'll add code to check the base class too.

File: tests/baselines/reference/bestCommonTypeOfTuple2.types (hunk 5 above)
Comment (sandersn) [inline tests/baselines/reference/bestCommonTypeOfTuple2.types]: After lots of discussion with you and @RyanCavanaugh, I propose a more complex rule that would revert this test change [1]:
1. If a property has a type annotation, use that.
2. If a property has an inherited type and **no** initialiser, use that.
3. If a property has an inherited type and an initialiser, contextually type the initialiser with the inherited type, then use the type of the initialiser.

This makes the following examples correct:

```ts
interface Comparer {
  compare(x: any, y: any): number;
}
class NumCmp implements Comparer {
  compare = (x: number, y: number) => x - y;
}
let cmp = new NumCmp();
cmp.compare("foo", "bar") // ERROR here

interface Long {
  length: number;
}
class Cat implements Long {
  length = undefined;
}
let longcat = new Cat();
longcat.length = "error here"; // ERROR
```

[1] Note that the behaviour that caused the change is technically correct but it is a breaking change.

Comment (sandersn) [inline tests/baselines/reference/bestCommonTypeOfTuple2.types]: After more discussion, we decided to drop rule (2) and additionally special-case contextual typing of `null`, `undefined`, `[]` and `{}` so that `undefined/null` in these types takes the contextual type rather than `undefined`.

Comment (DanielRosenwasser) [inline tests/baselines/reference/bestCommonTypeOfTuple2.types]: I think that seems reasonable; so that means a property with no initialized implicitly get the `any` type?

Comment (sandersn) [inline tests/baselines/reference/bestCommonTypeOfTuple2.types]: Yes, this is the case like
```ts
class Base {
  f(n: number): number;
}
class Derived extends Base {
  f(n);
}
```

Comment (JsonFreeman) [inline tests/baselines/reference/bestCommonTypeOfTuple2.types]: So n is any, and f returns any?

Comment (JsonFreeman) [inline tests/baselines/reference/bestCommonTypeOfTuple2.types]: What do you plan to do when inferentially typing undefined or null? Will you leave it alone, to be later added as an inference candidate if it aligns with a type parameter? Or will you fix the type parameter, and then use that inference result to contextually type the undefined?

Comment (sandersn) [inline tests/baselines/reference/bestCommonTypeOfTuple2.types]: Can you give an example? The case I can come up with isn't very interesting -- its behaviour wouldn't change because the contextual type wouldn't change anything.

```ts
let f: (n: number) => number;
let g: <T>(t: T) => void;
f(undefined); // first argument: number
f(undefined); // first argument: undefined which widens to any, so t is any.
```

Comment (DanielRosenwasser) [inline tests/baselines/reference/bestCommonTypeOfTuple2.types]: I think this is the example @JsonFreeman is thinking of:

```ts
declare function f<T>(x: T, y: T): T;

const a = f(100, undefined);
const b = f(undefined, 200);
```

What will happen here? What will we infer for the types of `a` and `b`?

Comment (JsonFreeman) [inline tests/baselines/reference/bestCommonTypeOfTuple2.types]: Yeah something like that is what I meant. In @DanielRosenwasser, a and b should both be number. So for null and undefined, the contextual type should only take effect if you are not in an inference context.

File: src/compiler/checker.ts [hunk 6]
```diff
@@ -2595,7 +2595,14 @@ namespace ts {
 
             // Use the type of the initializer expression if one is present
             if (declaration.initializer) {
-                return checkExpressionCached(declaration.initializer);
+                let mapper: TypeMapper;
+                if (declaration.kind === SyntaxKind.PropertyDeclaration) {
+                    const type = getTypeOfBasePropertyDeclaration(<PropertyDeclaration>declaration);
+                    if (type) {
+                        mapper = createTypeMapper([undefinedType, nullType], [type, type]);
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Since the first argument never changes, you could factor it out to a constant.

File: src/compiler/checker.ts [hunk 7]
```diff
@@ -2895,6 +2902,40 @@ namespace ts {
             return unknownType;
         }
 
+        function getTypeOfBasePropertyDeclaration(declaration: PropertyDeclaration) {
+            if (declaration.parent.kind === SyntaxKind.ClassDeclaration) {
+                const property = getPropertyOfBaseTypeDeclaration(<ClassLikeDeclaration>declaration.parent, declaration.symbol.name);
+                if (property) {
+                    return getTypeOfSymbol(property);
+                }
+            }
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Can you use an explicit return?

File: src/compiler/checker.ts [hunk 8]
```diff
@@ -2895,6 +2902,40 @@ namespace ts {
...
+
+        function getPropertyOfBaseTypeDeclaration(declaration: ClassLikeDeclaration, propertyName: string): Symbol {
+            const property = getFirstPropertyOfTypes(getBaseTypes(<InterfaceType>getTypeOfSymbol(getSymbolOfNode(declaration))), propertyName);
+            if (property) {
+                return property;
+            }
+            const implementedTypeNodes = getClassImplementsHeritageClauseElements(declaration);
+            if (implementedTypeNodes) {
+                return getFirstPropertyOfTypes(map(implementedTypeNodes, getTypeFromTypeReference), propertyName);
+            }
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Same here; explicit return

File: src/compiler/checker.ts (hunk 6 above)
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: I'm not so sure this works -  doesn't this map *all* undefined and null types in the expression? So if I have

```ts
class Base {
    a: (x: string) => void;
}

class Derived extends Base {
    a = (x = null) => {}
}
```

won't `x` in `Derived` get the type `(x: string) => void` meaning `a` will have the type `(x: (x: string) => void) => void`.

Maybe I misunderstood.

Comment (JsonFreeman) [inline src/compiler/checker.ts]: +1

Also, I think the TypeMappers are supposed to be for generics.

File: src/compiler/checker.ts [hunk 9]
```diff
@@ -6925,7 +6966,11 @@ namespace ts {
             checkCollisionWithCapturedThisVariable(node, node);
             checkBlockScopedBindingCapturedInLoop(node, symbol);
 
-            return getNarrowedTypeOfSymbol(getExportSymbolOfValueSymbolIfExported(symbol), node);
+            const type = getNarrowedTypeOfSymbol(getExportSymbolOfValueSymbolIfExported(symbol), node);
+            if (type === undefinedType || type == nullType) {
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: Why is this here?

File: src/compiler/checker.ts [hunk 10]
```diff
@@ -2595,7 +2595,14 @@ namespace ts {
...
             if (declaration.initializer) {
-                return checkExpressionCached(declaration.initializer);
+                let mapper: TypeMapper;
+                if (declaration.kind === SyntaxKind.PropertyDeclaration) {
+                    const type = getTypeOfBasePropertyDeclaration(<PropertyDeclaration>declaration);
+                    if (type) {
+                        mapper = createTypeMapper([undefinedType, nullType], [type, type]);
+                    }
+                }
+                return checkExpressionCached(declaration.initializer, mapper);
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: The `contextualMapper` parameter isn't intended to be used this way. The comment on `checkExpression` explains its purpose.

File: src/compiler/checker.ts [hunk 11]
```diff
@@ -6925,7 +6966,11 @@ namespace ts {
             checkCollisionWithCapturedThisVariable(node, node);
             checkBlockScopedBindingCapturedInLoop(node, symbol);
 
-            return getNarrowedTypeOfSymbol(getExportSymbolOfValueSymbolIfExported(symbol), node);
+            const type = getNarrowedTypeOfSymbol(getExportSymbolOfValueSymbolIfExported(symbol), node);
+            if (type === undefinedType || type == nullType) {
+                return (contextualMapper || identityMapper)(type);
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: The `contextualMapper` parameter doesn't indicate whether you have a contextual type. You need to actually get the contextual type using `getContextualType`.

Comment (JsonFreeman) [inline src/compiler/checker.ts]: @ahejlsberg can correct me, but I think the contextualMapper is just for inferential typing.

**Comment by sandersn** [comment]
Yes, Anders explained contextual typing in person and I'll push new changes with `getContextualType` as soon as I make sure that all the baseline changes are correct.

File: src/compiler/checker.ts (hunk 9 above)
Comment (sandersn) [inline src/compiler/checker.ts]: This checks for the undefined type instead of the symbol with the name "undefined". I think this is wrong, so I changed it to `if (symbol.name === "undefined")`

File: src/compiler/checker.ts (hunk 7 above)
Comment (sandersn) [inline src/compiler/checker.ts]: done

File: src/compiler/checker.ts (hunk 8 above)
Comment (sandersn) [inline src/compiler/checker.ts]: done

**Comment by sandersn** [comment]
Here's why `wideningTuples3` fails to report a noImplicitAny error:

```ts
var a: any[];
var b = a = [undefined, null];
```

`a` contextually types `[undefined, null]: any[]`, which in turn contextually types `undefined` and `null` as `any`. Then `b` infers the type `[any, any]` since it has an array of two elements, each of type `any`.

But this doesn't trigger the implicit any check in `reportErrorsFromWidening` because it looks for types that *will be* widened to `any` -- `undefined` or `null`. But these particular undefined/null values have already been assigned the type `any`. It's arguably no longer implicit since their types were contextually propagated from the explicit `var a: any[]`.

I'm not sure whether this needs to be fixed, or whether it's an obscure enough corner case to let the technicality slide.

**Comment by sandersn** [comment]
After discussing with Mohamed, I think dropping the implicit any error is the right thing in this case.

**Comment by RyanCavanaugh** [comment]
Observable types from contextually-typed expressions have always been very weird, so I don't think it's much to worry about. I think that code has changed meaning more than once since 1.0.

**Comment by JsonFreeman** [comment]
I agree with @mhegazy and @sandersn about dropping the no implicit any error.

**Comment by DanielRosenwasser** [comment]
Remind me, does this also fix #1373?

**Comment by sandersn** [comment]
Yes, in the below code, `x` is now contextually typed as `number`.
```ts
interface I {
 method(n: number): number
}
class C implements I {
  method = x => x + 1
}
```

**Comment by sandersn** [comment]
@DanielRosenwasser, I added the test case from #6772 and fixed the bug that it exposed.

**Comment by ahejlsberg** [comment]
Should we also contextually type method parameters as suggested in #6854?

**Comment by sandersn** [comment]
055ae8c adds method contextual typing too. See #6854 for an excerpt from the tests added in that commit.

**Comment by sandersn** [comment]
Two notes from discussion with @DanielRosenwasser :

1. Add a test case for multiple interfaces:

```ts
interface X {
  n: string;
}
interface Y {
  n: number;
}
class C implements X, Y {
  n = undefined; // n should be number & string
}
```

2. Non-nullable types should behave this way once they are in:

```ts
interface X {
  n: string; // n cannot be null
}
class C implements X {
  n = undefined; // contextually type it as (string | undefined), which should then error.
}
```

**Comment by sandersn** [comment]
Another test case with string literals:

```ts
class Foo {
    bar: "fizz" | "buzz";
}

class Bar extends Foo {
    bar = "fizz";
}
```

Without this change, it fails because types of bar are not compatible: 'string' is not assignable to '"fizz" | "buzz"'.

**Comment by DanielRosenwasser** [comment]
@sandersn it "fixes it" but what is the behavior? Does `Bar`'s `bar` property have type `"fizz"` or "fizz" | "buzz"`?

I would argue that the latter is more desirable because it keeps things open ended, but I might be wrong. It would make "branding" a subtype harder. For instance, you'd need to write

```ts
class Bar {
    bar: "fizz" = "fizz";
}
```

**Comment by sandersn** [comment]
Right now the contextual type rules for string literals means that 
`bar = "fizz"` produces `bar: "fizz"`. Which I agree is not as good as `bar: "fizz" | "buzz"`

But the contextual typing of a string literal is an orthogonal problem to whether properties are contextually typed.

**Comment by sandersn** [comment]
# Running on real world code

We discussed this PR in the design meeting last week and decided that it was too inconsistent and didn't capture intent well enough. Instead, I tried a simpler approach that gives inherited properties *exactly* their inherited type (unless there is a type annotation). The change is found in the branch contextually-type-inherited-properties-WIP. I ran this branch on our real world code base consisting of Microsoft internal code as well as projects like Angular and Grunt. This showed that the new change broke more than it helped. Since both approaches failed, I am going to close this PR -- we may revisit this problem in the future.


## The good

The improved typing exposed one bug that happens twice in the same project:

```ts
interface I {
  method(o: SomeType): string;
}
class C extends I {
  method(o) {
    return o.doesNotExist;
  }
}
```

Before this change, `o` was implicitly `any` in the parameter list of `C#method`.
Now an error is reported because `o: SomeType`, which does not have a
property named `doesNotExist`.

Both methods would have caught this bug. So would `--noImplicitAny`.

## The bad

Consider the following example:

```ts
namespace Base {
    export interface Item {
        a;
    }
    export abstract class Base {
        abstract item: Item;
    }
}
namespace User {
    export interface Subitem extends Base.Item {
        sub;
    }
    export interface Contract {
        item: Subitem;
    }
    export function createC(): Contract {
    }
    function createSubitem(): Subitem {
        // factory code here ...
    }
    class C extends Base.Base implements Contract {
        item = createSubitem();
    }
}
```

The intended usage is something like

```ts
let c: User.Contract = User.createC();
console.log(c.item.a);
console.log(c.item.sub);
```

Unfortunately, if `C.item` takes its type from `Base.item`, then
`C.item: Item`. Then `C` no longer implements `Contract` because
`Contract.item: Subitem`.

## The ugly

In several projects, properties of type KnockoutObservable changed to
KnockoutObservableBase, which is a superclass that has fewer
properties. This can be fixed by upgrading to a non-ancient version of
knockout.d.ts, though. Current versions of knockout.d.ts don't have
KnockoutObservableBase.

**Comment by DanielRosenwasser** [comment]
👍  Thanks for reporting on what sorts of problems/improvements this brought.

**Comment by tinganho** [comment]
if `A` extends `B` and implements `C`. Why doesn't it just contextually type based on members of `B` and `C` and when there exists common members contextually type based on the most sub subtype member of `B` and `C`?

Wouldn't that remove the bad and ugly part?

**Comment by JsonFreeman** [comment]
@tinganho I believe that in general there is not a lowest subtype for a set of types. The type system does not provide a total ordering.

**Comment by DanielRosenwasser** [comment]
Keywords: base derived methods properties property contextual contextually type subclass superclass
